torch/_inductor/scheduler.py

```diff
@@ -850,6 +943,11 @@ def check_all_pairs(nodes):
         for node_grouping in buffer_names_grouping.values():
             check_all_pairs(node_grouping)
 
+        if config.fusion_reorder_loops:
```

Comment:
Any particular reason it's structured this way rather than modifying `can_fuse`?

Reply:
I like the fact that the reordering logic is self-contained within this one function and `can_fuse` remains relatively easy to read and understand. It also does the reordering at the ir level, so is backend agnostic.

Reply:
IMO we should put this in `can_fuse()`.  I think it will simplify the code (no need to iterate over O(n^2) possible fusions twice), and also be faster since I suspect reordering will be slow since it calls FX symbolic trace.

I think you can still make it self contained, basically just change:
https://github.com/pytorch/pytorch/blob/e7f1daa355064fac64e2ac9cec5d51181549ab26/torch/_inductor/scheduler.py#L1209-L1214
to:
```
if remaining_deps & node1_names: 
    return self.can_fuse_via_reordering()
```

You then need to remember to apply the reordering when the actual fusion happens.  You could do something like:
```
self._fusion_needs_reordering[(node1,node2)] = ...
```

Reply:
> I suspect reordering will be slow since it calls FX symbolic trace.

If compilation latency is an issue I think we can re-use the original graph since in theory only the indexing formulas should change. 

> I think you can still make it self contained, basically just change: [Scheduler.can_fuse_vertical]

You also need to add support for reordering in:
- `get_backend(device).can_fuse_vertical` for all backends
- `score_fusion_memory` which is called in `can_fuse` before `can_fuse_vertical`
- `fuse_nodes_once`
 
This is not self-contained because it leaks all over the code and will make composing future transformations or adding new fusion scoring metrics significantly harder.